statemachine/statemachine.go

```diff
@@ -0,0 +1,317 @@
+package statemachine
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/lytics/metafora"
+)
+
+type StateCode string
+
+const (
+	Runnable  StateCode = "runnable"  // Scheduled
+	Sleeping            = "sleeping"  // Scheduled, not running until time has elapsed
+	Completed           = "completed" // Terminal, not scheduled
+	Killed              = "killed"    // Terminal, not scheduled
+	Failed              = "failed"    // Terminal, not scheduled
+	Fault               = "fault"     // Scheduled, in error handling / retry logic
+	Paused              = "paused"    // Scheduled, not running
+)
+
+// Terminal states will never run and cannot transition to a non-terminal
+// state.
+func (s StateCode) Terminal() bool {
+	switch s {
+	case Runnable, Sleeping, Paused, Fault:
+		return false
+	case Completed, Killed, Failed:
+		return true
+	default:
+		panic("unknown state " + s)
+	}
+}
+
+// State represents the current state of a stateful handler. See StateCode for
+// details. Until and Errors are extra state used by the Sleeping and Fault
+// states respectively.
+type State struct {
+	Code   StateCode  `json:"state"`
+	Until  *time.Time `json:"until,omitempty"`
+	Errors []Err      `json:"errors,omitempty"`
+}
+
+func (s State) String() string {
+	switch s.Code {
+	case Sleeping:
+		return fmt.Sprintf("%s until %s", s.Code, s.Until)
+	default:
+		return string(s.Code)
+	}
+}
+
+// Messages are events that cause state transitions. Until and Err are used by
+// the Sleep and Error messages respectively.
+type Message struct {
+	Code MessageCode `json:"message"`
+
+	// Until is when the statemachine should transition from sleeping to runnable
+	Until *time.Time `json:"until,omitempty"`
+
+	// Err is the error that caused this Error message
+	Err error `json:"error,omitempty"`
+}
+
+// MessageCode is the symbolic name of a state transition.
+type MessageCode string
+
+const (
```

Comment:
By having these be `const` it means implementers can't add their own states right?    Should we make this more flexable like aaron did with QLbridge's functions `https://github.com/araddon/qlbridge/blob/9d9714a5f29881090620558bf0b9e6f7cd77f8ac/expr/builtins/builtins.go#L21` 

The rules for transition are already flexible, so no further commented needed there.